### DISCUSSION TITLE
Joining Multiple Records in Query

### DISCUSSION THREAD

**Discussion by no1melman** [ravendb/ravendb#14889] [Q&A] [answered]
I've got this setup

```fsharp
type Role = { Id: string; Name: string }

type User = { Id: string; Name: string; Roles: Role list }
```

I want to get all the roles back, so

```fsharp
session.Query<Role>().Join(
            foundUser.Roles.AsEnumerable(),
            (fun r -> r.Id.Get),
            (fun r -> r.Id.Get),
            (fun foundRole userRole -> foundRole))
         .ToListAsync(token = ctx.RequestAborted)
```

I get this back

```text
An unhandled exception has occurred while executing the request.
System.NotSupportedException: Could not understand expression: from 'Roles'.Join([{ Id = GeneralPurposeId "roles/1-A"
  Name = "Basic" }; { Id = GeneralPurposeId "roles/2-A"
  Name = "Power" }; { Id = GeneralPurposeId "roles/3-A"
  Name = "Administrator" }], r => r.Id.Get, r => r.Id.Get, (foundRole, userRole) => foundRole)
 ---> System.NotSupportedException: Method not supported: Join
```

Is this a problem with my join code - or is this that joins aren't supported?

**Comment by ayende** [comment, marked as answer]
Joins as you are trying to do them are not supported.
However, RavenDB has the `Include` feature, which will look ( in C#) something like this:

Going from the users by the role: `session.Query<User>().Where(x=>x.Roles.In(foundRole)).ToList()`
Going from the user and the roles: `session.Query<User>().Where(x=>x.Name == userName).Include(x=>x.Roles).ToList()`

In the first case, you are just getting the matching users for the role (note, this may be a *very* big list, there aren't many calls to do that).
In the second, you are getting the users and its roles in a single remote calls. Future calls to `session.Load(role)` will be served from the session cache directly, no network calls.


See full discussion here: https://ravendb.net/docs/article-page/5.4/Csharp/client-api/how-to/handle-document-relationships#includes
